### PR TITLE
Feature/make aql analyzer type aware

### PR DESCRIPTION
**Pull request by Dronplane**
### Scope & Purpose

Add new property for Aql analyzer  - returnType.
This allows to convert emitted by query values to strings, numbers or booleans and automatically apply internal numeric/boolean analyzers for tokenizing corresponding values.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**
  - [x] Added new **integration tests**  shell_server/client_aql
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### CODE REVIEW

File: arangod/IResearch/IResearchAqlAnalyzer.cpp [hunk 1]
```diff
@@ -486,13 +527,63 @@ bool AqlAnalyzer::next() {
...
-        if (value.isString() || (value.isNull(true) && _options.keepNull)) {
-          if (value.isString()) {
-            _term.value = arangodb::iresearch::getBytesRef(value.slice());
-          } else {
-            _term.value = irs::bytes_ref::EMPTY;
+        if (value.isString() ||
+            (_options.keepNull && value.isNull(true) ) ||
+            (_options.returnType != AnalyzerValueType::Undefined && !value.isNull(true))) {
+          switch (_options.returnType) {
+            case AnalyzerValueType::Undefined:
```
Comment (gnusi) [inline arangod/IResearch/IResearchAqlAnalyzer.cpp:534]: I suggest we get rid of  `AnalyzerValueType::Undefined` as it only adds extra complexity without providing any benefits

File: arangod/IResearch/IResearchAqlAnalyzer.h [hunk 2]
```diff
@@ -43,9 +50,19 @@ class AqlAnalyzer final : public irs::analysis::analyzer{
...
+      keepNull(keep), batchSize(batch), memoryLimit(limit), returnType(retType) {}
+
+    bool substreamInUse() const noexcept {
+      return returnType == AnalyzerValueType::Bool ||
+             returnType == AnalyzerValueType::Number;
+    }
+
+    bool substreamWillCountPositions() const noexcept {
+      return substreamInUse() && !collapsePositions;
+    }
```
Comment (gnusi) [inline arangod/IResearch/IResearchAqlAnalyzer.h:65]: It seems we don't need these functions anymore

File: arangod/IResearch/IResearchAqlAnalyzer.h [hunk 3]
```diff
@@ -66,6 +83,12 @@ class AqlAnalyzer final : public irs::analysis::analyzer{
 
     /// @brief memory limit for query.  1Mb by default. Could be increased to 32Mb
     uint32_t memoryLimit{ 1048576U };
+
+    /// @brief target type to convert query output. Could be
+    ///        string, bool, number. If missing - backward compatible mode is used
+    ///        strings are passed verbatim, nulls discarder/replaced other types
+    ///        silently discarded.
+    AnalyzerValueType returnType{AnalyzerValueType::Undefined};
```
Comment (gnusi) [inline arangod/IResearch/IResearchAqlAnalyzer.h:91]: string by default?

File: arangod/IResearch/IResearchAqlAnalyzer.h [hunk 4]
```diff
@@ -79,31 +102,35 @@ class AqlAnalyzer final : public irs::analysis::analyzer{
...
-  irs::increment _inc;
-  std::string _str;
+  using attributes = std::tuple<
+    irs::increment,
+    AnalyzerValueTypeAttribute,
+    irs::term_attribute,
+    VPackTermAttribute>;
+
   Options _options;
+  arangodb::aql::AqlValue _valueBuffer;
```
Comment (gnusi) [inline arangod/IResearch/IResearchAqlAnalyzer.h:120]: we can shorten strings as we're already in `arangodb`  namespace

File: arangod/IResearch/IResearchDocument.h [hunk 5]
```diff
@@ -196,7 +199,14 @@ class FieldIterator {
   arangodb::transaction::Methods* _trx;
   irs::string_ref _collection;
   Field _value;  // iterator's value
-  IndexId _linkId;    
+  IndexId _linkId;
+
+  // Support for outputting primitive type from analyzer
+  using PrimitiveTypeResetter = void (*)(irs::token_stream* stream, VPackSlice slice);
```
Comment (gnusi) [inline arangod/IResearch/IResearchDocument.h:205]: let's move this alias above members definition.

**Review by gnusi — CHANGES_REQUESTED** [review]

File: arangod/IResearch/IResearchAqlAnalyzer.cpp (hunk 1 above)
Comment (Dronplane) [reply arangod/IResearch/IResearchAqlAnalyzer.cpp:534]: As we already released documentation I wanted to retain that behaviour: no implicit conversion (as setting returnType:'string' will enable conversion). But as we have not released final version of binaries we could drop support and declare "always implicit conversion to string by default". WDYT?

File: arangod/IResearch/IResearchAqlAnalyzer.h (hunk 3 above)
Comment (Dronplane) [reply arangod/IResearch/IResearchAqlAnalyzer.h:91]: If we decide to drop old behaviour (see comment above) than yes.

**Review by gnusi — APPROVED** [review]

**Pull request merged**